components/core/src/event.rs

```diff
@@ -0,0 +1,139 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::fs::{self, File};
+use std::io::{Write, Read};
+use std::path::Path;
+use std::time::{UNIX_EPOCH, SystemTime};
+use uuid::Uuid;
+use rustc_serialize::json::{ToJson, Json};
+use fs::cache_analytics_path;
+
+// Supported events
+pub const EVENT_BUILDER_PROJECT_CREATE: &'static str = "builder-project-create";
+
+// Sample event JSON payload (compatible with Segment.io)
+// {
+//   "type": "track",
+//   "event": "builder-project-create",
+//   "properties": {
+//     "clientid" : "0a5c0882-ade5-46cf-821d-8d3853cd0d41"
+//     "timestamp": "1479330000.13442404",
+//   }
+// }
+
+const CLIENT_ID_METAFILE: &'static str = "CLIENT_ID";
+
+#[derive(Debug, Clone)]
+pub struct Event {
+    name: String,
+    clientid: String,
+    timestamp: String,
+    properties: BTreeMap<String, String>,
+}
+
+impl Event {
+    pub fn new(name: &str, clientid: &str, timestamp: &str) -> Self {
+        let mut properties = BTreeMap::new();
+        properties.insert("timestamp".to_string(), timestamp.to_string());
+        properties.insert("clientid".to_string(), clientid.to_string());
+
+        Event {
+            name: name.to_string(),
+            clientid: clientid.to_string(),
+            timestamp: timestamp.to_string(),
+            properties: properties,
+        }
+    }
+}
+
+impl ToJson for Event {
+    fn to_json(&self) -> Json {
+        let mut p = BTreeMap::new();
+        for (key, value) in self.properties.iter() {
+            p.insert(key.to_string(), value.to_json());
+        }
+
+        let mut m = BTreeMap::new();
+        m.insert("type".to_string(), "track".to_string().to_json());
+        m.insert("event".to_string(), self.name.to_json());
+        m.insert("properties".to_string(), p.to_json());
+
+        Json::Object(m)
+    }
+}
+
+fn read_file(file_path: &Path) -> String {
+    let mut content = String::new();
+    let mut file = File::open(file_path).expect("Unable to open file");
+    file.read_to_string(&mut content).expect("Unable to read file");
+    content
+}
+
+fn write_file(parent_dir: &Path, file_path: &Path, content: &str) {
+    fs::create_dir_all(parent_dir).expect("Unable to create directory");
+    let mut file = File::create(&file_path).expect("Unable to create file");
+    file.write_all(content.as_bytes()).expect("Unable to write file");
+}
+
+fn timestamp() -> String {
+    let (secs, subsec_nanos) = match SystemTime::now().duration_since(UNIX_EPOCH) {
+        Ok(duration) => (duration.as_secs(), duration.subsec_nanos()),
+        Err(e) => {
+            debug!("Cannot generate system time: {}", e);
```

Comment:
This should be `error!()` and not `debug!()`. We want the service to keep going even if it's generating less than ideal data, but we want our monitoring tools to pick it up and alert us!